mam/v2/api/api.c

```diff
@@ -183,12 +198,94 @@ retcode_t mam_api_bundle_write_packet(mam_api_t *const api,
   return RC_OK;
 }
 
-retcode_t mam_api_bundle_read(mam_api_t *const api,
-                              mam_channel_t const *const cha,
-                              bundle_transactions_t const *const bundle,
-                              flex_trit_t *const payload) {
-  if (api == NULL || cha == NULL || bundle == NULL || payload == NULL) {
-    return RC_NULL_PARAM;
+bool mam_api_bundle_contains_header(bundle_transactions_t const *const bundle) {
+  iota_transaction_t *tx = (iota_transaction_t *)utarray_front(bundle);
+  return flex_trits_are_null(transaction_tag(tx), FLEX_TRIT_SIZE_81);
+}
+
+trits_t msg_trits_from_bundle(bundle_transactions_t const *const bundle,
+                              const trit_t msg_trits[],
+                              size_t num_trits_in_bundle, size_t start_index) {
+  iota_transaction_t *curr_tx = (iota_transaction_t *)utarray_eltptr(bundle, 0);
+
+  trit_t *msg_trits_ptr = msg_trits;
+  size_t curr_tx_index = 0;
+  BUNDLE_FOREACH(bundle, curr_tx) {
+    if (curr_tx_index++ < start_index) {
+      continue;
+    }
+    flex_trits_to_trits(msg_trits_ptr, NUM_TRITS_SIGNATURE,
+                        transaction_signature(curr_tx), NUM_TRITS_SIGNATURE,
+                        NUM_TRITS_SIGNATURE);
+    msg_trits_ptr += NUM_TRITS_SIGNATURE;
   }
-  return RC_OK;
+
+  return trits_from_rep(num_trits_in_bundle, msg_trits);
+}
+
+static retcode_t mam_api_bundle_read_header(
+    mam_msg_recv_context_t const *const ctx, trits_t *const msg) {
+  retcode_t err;
+  ERR_BIND_RETURN(mam_msg_recv(ctx, msg), err);
+  return err;
+}
+
+retcode_t mam_api_bundle_read_msg(mam_api_t *const api,
+                                  flex_trit_t const *const ch_id,
+                                  bundle_transactions_t const *const bundle,
+                                  flex_trit_t **const packet_payload) {
+  retcode_t err = RC_OK;
+  MAM2_ASSERT(packet_payload && *packet_payload == NULL);
+  if (!mam_api_bundle_contains_header(bundle)) {
```

Comment:
Why do we return if there is no header ? That won't work on single packets

Reply:
Of course it won't, single packets are going to be read with `mam_api_bundle_read_packet`